fedimint-api-client/src/api/mod.rs

```diff
@@ -919,123 +968,261 @@ impl WsFederationApi<WsClient> {
-            // if we're looping again and `attempts` is greater than `RETRIES`, we have a
-            // bug.
-            debug_assert!(attempts <= RETRIES);
-
-            let rclient = self.client.read().await;
-            match rclient.client.get_try().await {
-                Ok(client) if client.is_connected() => {
-                    return client.request::<_, _>(method, params).await;
-                }
-                Err(err) => {
-                    if RETRIES <= attempts {
-                        return Err(JsonRpcClientError::Transport(err.into()));
-                    }
-                    debug!(target: LOG_CLIENT_NET_API, err=%err.fmt_compact(), "Triggering reconnection after connection error");
-                }
-                Ok(_client) => {
-                    if RETRIES <= attempts {
-                        return Err(JsonRpcClientError::Transport(
-                            anyhow::format_err!("Disconnected").into(),
-                        ));
+#[derive(Clone, Debug)]
+struct ClientConnection {
+    sender: async_channel::Sender<oneshot::Sender<DynClientConnection>>,
+}
+
+impl ClientConnection {
+    fn new(peer: PeerId, connector: DynClientConnector) -> ClientConnection {
+        let (sender, receiver) = bounded::<oneshot::Sender<DynClientConnection>>(128);
+
+        fedimint_core::task::spawn(
+            "peer-api-connection",
+            async move {
+                let mut backoff = api_networking_backoff();
+
+                while let Ok(sender) = receiver.recv().await {
+                    match connector.connect(peer).await {
+                        Ok(connection) => {
+                            info!(target: LOG_CLIENT_NET_API, "Connected to peer api");
+
+                            sender.send(connection.clone()).ok();
+
+                            loop {
+                                tokio::select! {
+                                    sender = receiver.recv() => {
+                                        match sender.ok() {
+                                            Some(sender) => sender.send(connection.clone()).ok(),
+                                            None => break,
+                                        };
+                                    }
+                                    () = connection.await_disconnection() => break,
+                                }
+                            }
+
+                            info!(target: LOG_CLIENT_NET_API, "Disconnected from peer api");
+
+                            backoff = api_networking_backoff();
+                        }
+                        Err(e) => {
+                            info!(target: LOG_CLIENT_NET_API, "Failed to connect to peer api {e}");
+
+                            // We need to drain the channel for the pending requests to fail.
+
+                            while receiver.try_recv().is_ok() {}
```

Comment:
I very much doubt you'd hit that in the CI, but draining after error, is technically still a race condition, because the requester could have arrived after connection was made, but before the queue was drained.

The queue needs to be drained before `.connector.connect` and if connect succeeded, all requesters receive their connector (otherwise they can get dropped).